straxen/common.py

```diff
@@ -243,7 +247,185 @@ def get_livetime_sec(context, run_id, things=None):
+    # for old runs.
+    remap = get_resource(
+        aux_repo + '/ecb6da7bd4deb98cd0a4e83b3da81c1e67505b16/remapped_channels_since_20200729_17.20UTC.csv',
+        fmt='csv')
+
+    def wr_tqdm(x):
+        """Wrap input x with tqdm"""
+        if _tqdm:
+            try:
+                return tqdm.tqdm_notebook(x)
+            except (AttributeError, ModuleNotFoundError, ImportError):
+                # ok, sorry lets not wrap but return x
+                pass
+        return x
+
+    def get_dtypes(_data):
+        """
+        Return keys/dtype names of pd.DataFrame or numpy array
+
+        :param _data: data to get the keys/dtype names
+
+        :return: keys/dtype names
+        """
+        if isinstance(_data, np.ndarray):
+            _k = _data.dtype.names
+        elif isinstance(_data, pd.DataFrame):
+            _k = _data.keys()
+        return _k
+
+    def convert_channel(_data, replace=('channel', 'max_pmt')):
+        """
+        Given an array, replace the 'channel' entry if we had to remap it according to the
+            map of channels to be remapped
+
+        :param _data: data whereof the replace entries should be changed according to the
+            remapping of channels
+
+        :param replace: entries (keys/numpy-dtypes) that should be changed in the data
+
+        :return: remapped data where each of the replace entries has been replaced
+        """
+        # loop over the things to replace
+        for _rep in replace:
+            _k = get_dtypes(_data)
+            if _rep not in _k:
+                # Apparently this data doesn't have the entry we want to replace
+                continue
+            # Make a buffer we can overwrite and replace with an remapped array
+            buff = np.array(_data[_rep])
+            _n_replacements = 0
+            for _, _row in wr_tqdm(remap.iterrows()):
+                pmt_new, pmt_old = _row['PMT_new'], _row['PMT_old']
+                mask = _data[_rep] == pmt_old
+                buff[mask] = pmt_new
+                _n_replacements += np.sum(mask)
+            _data[_rep] = buff
+            if verbose:
+                print(f'convert_channel::\tchanged {_n_replacements}/{len(_data)} items')
+        # Not needed for np.array as the internal memory already reflects it but it is
+        # needed for pd.DataFrames.
+        return _data
+
+    def remap_single_entry(_data, _entry):
+        """
+        Remap the data of a single entry.
+
+        :param _data: reshuffle the _data according to for _entry according to the map of
+            channels to be remapped
+
+        :param _entry: key or dtype of the data.
+
+        :return: correctly mapped data
+        """
+        _k = get_dtypes(_data)
+        if _entry not in _k:
+            raise ValueError(f'remap_single_entry::\tcannot remap {_entry} in data with'
+                             f' fields {_k}')
+        buff = np.array(_data[_entry])
+        for _, _row in remap.iterrows():
```

Comment:
Do you have to loop here. Can you not do simply new[:, [1, 2, 3, 4, 5]] = old[:, [6, 7, 8, 9, 10]], might be faster.

Reply:
Looks more elegant what you propose anyway, let me check

Reply:
interestingly enough,  it is the fasted.
```python
%%timeit
remap_single_entry(peaks, 'area_per_channel')
```
The implementation as above:
1.84 s ± 44.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
What you propose (``buff[:, np.array(remap['PMT_new'].values)] = _data[_array_entry][:, np.array(remap['PMT_old'].values)]``)
2 s ± 61.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Numba (first run once to initialize, after that):
2.78 s ± 46.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

Reply:
That is indeed interesting. Looks a bit odd to me that numba is slower here. I mean the function would be the same except that the for-loop should get a boost.

Reply:
one thing to keep in mind is that there are only ~20 items in this for loop. The looping here might go faster but is a rather small part of the computation time